backend/src/BuiltinExecution/Libs/String.fs

```diff
@@ -299,20 +299,44 @@ let fns : List<BuiltInFn> =
           Param.make "to" TInt "" ]
       returnType = TString
       description =
-        "Returns the substring of <param string> between the <param from> and <param
-         to> indices.
-
+        "Returns the substring of <param string> between the <param from> and <param to> indices.
          Negative indices start counting from the end of <param string>."
       fn =
         (function
         | _, _, [ DString s; DInt first; DInt last ] ->
-          String.toEgcSeq s
-          |> Seq.toList
-          |> FSharpPlus.List.drop (int first)
-          |> List.truncate (int (last - first))
-          |> String.concat ""
-          |> DString
-          |> Ply
+          let strLength = String.length s
+          // Adjust first and last indices if they are negative (to start counting from the end)
+          let first = if first < 0 then strLength + int first else int first
+          let last = if last < 0 then strLength + int last else int last
+
+          // Return empty string if first index is out of bounds or greate than or equal to the last index
+          if first >= strLength || first >= last then
+            Ply(DString "")
+          else
+            // Create a TextElementEnumerator to iterate through the string elements
+            // This is used to handle EGCs correctly
+            let textElemEnumerator =
+              System.Globalization.StringInfo.GetTextElementEnumerator(s)
+            let mutable startIndex = 0
+            let mutable endIndex = 0
+            let mutable index = 0
+
+            // Iterate through the string elements using the enumerator
```

Comment:
This comments explains exactly what is happening, but it doesn't tell us any more than the code does.

Maybe "Iterate through EGCs and record the byte indexes of the start and end"